exoscale/resource_exoscale_nlb.go

```diff
@@ -0,0 +1,275 @@
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutUpdate))
+	defer cancel()
+
+	client := GetComputeClient(meta)
+
+	nlb, err := findNLB(ctx, d, meta)
+	if err != nil {
+		return err
+	}
+
+	if nlb == nil {
+		return fmt.Errorf("Network Load Balancer %q not found", d.Id())
+	}
+
+	if d.HasChange("name") {
+		nlb.Name = d.Get("name").(string)
+	}
+
+	if d.HasChange("description") {
+		nlb.Description = d.Get("description").(string)
+	}
+
+	zone := d.Get("zone").(string)
+
+	ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), zone))
+	_, err = client.UpdateNetworkLoadBalancer(ctx, zone, nlb)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] %s: update finished successfully", resourceNLBIDString(d))
+
+	return resourceNLBRead(d, meta)
+}
+
+func resourceNLBDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: beginning delete", resourceNLBIDString(d))
+
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout(schema.TimeoutDelete))
+	defer cancel()
+
+	client := GetComputeClient(meta)
+
+	zone := d.Get("zone").(string)
+
+	ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), zone))
+	err := client.DeleteNetworkLoadBalancer(ctx, zone, d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] %s: delete finished successfully", resourceNLBIDString(d))
+
+	return nil
+}
+
+func resourceNLBApply(ctx context.Context, client *egoscale.Client, d *schema.ResourceData, nlb *egoscale.NetworkLoadBalancer) error {
+	if err := d.Set("name", nlb.Name); err != nil {
+		return err
+	}
+
+	if err := d.Set("description", nlb.Description); err != nil {
+		return err
+	}
+
+	if err := d.Set("created_at", nlb.CreatedAt.String()); err != nil {
+		return err
+	}
+
+	if err := d.Set("state", nlb.State); err != nil {
+		return err
+	}
+
+	if err := d.Set("ip_address", nlb.IPAddress.String()); err != nil {
+		return err
+	}
+
+	services := make([]string, len(nlb.Services))
+	for i, service := range nlb.Services {
+		services[i] = service.Name
+	}
+
+	if err := d.Set("services", services); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func findNLB(ctx context.Context, d *schema.ResourceData, meta interface{}) (*egoscale.NetworkLoadBalancer, error) {
+	client := GetComputeClient(meta)
+
+	if zone, ok := d.GetOk("zone"); ok {
+		ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), zone.(string)))
+		nlb, err := client.GetNetworkLoadBalancer(ctx, zone.(string), d.Id())
+		if err != nil {
+			return nil, err
+		}
+
+		return nlb, nil
+	}
+
+	ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), ""))
+	resp, err := client.RequestWithContext(ctx, egoscale.ListZones{})
+	if err != nil {
+		return nil, err
+	}
+	zones := resp.(*egoscale.ListZonesResponse).Zone
+
+	var nlb *egoscale.NetworkLoadBalancer
+	for _, zone := range zones {
+		n, err := client.GetNetworkLoadBalancer(ctx, zone.Name, d.Id())
```

Comment:
Slight change to the implementation: the current zone has to be specified for each iteration:

```suggestion
	resp, err := client.RequestWithContext(ctx, egoscale.ListZones{})
	if err != nil {
		return nil, err
	}
	zones := resp.(*egoscale.ListZonesResponse).Zone

	var nlb *egoscale.NetworkLoadBalancer
	for _, zone := range zones {
		n, err := client.GetNetworkLoadBalancer(
		apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), zone.Name)),
		zone.Name,
		d.Id())
```